Documentation/planning/multi-sdk-band-support.md

```diff
@@ -0,0 +1,258 @@
+5. **Checkout** 2xx branch
+6. **Build** using `./build.sh --with-sdk <PSB 1xx SDK> --with-artifacts <2xx inputs>`
+
+If a 2xx SDK is required to build the initial 2xx SDK, then the following workflow is used:
+
+1. **Checkout** 1xx branch:
+2. **Build** using `./build.sh --with-sdk <PSB 1xx SDK> --with-artifacts <PSB 1xx artifacts>`.
+3. **Gather** 1xx build outputs from `artifacts/x64/Release` = `<CSB 1xx artifacts>`.
+4. **Prepare** 2xx bootstrap inputs - `<PSB 1xx artifacts> + <CSB 1xx artifacts>`
+5. **Checkout** 2xx branch
+6. **Build** using `./build.sh --with-artifacts <2xx bootstrap inputs>`
+7. **Prepare** 2xx inputs - `<PSB 1xx artifacts> + <CSB 1xx artifacts> + <CSB 2xx artifacts>`
+8. **Build** using `./build.sh --with-sdk <CSB 2xx SDK> --with-artifacts <2xx inputs>`
+
+### A distro maintainer wishing to support only a 2xx SDK (monthly workflow)
+
+A distro maintainer who only wants to support the 2xx SDK still needs to build the 1xx branch to obtain the runtime and an updated 1xx SDK for use the next month. In this case, we build the 2xx inputs with the 2xx SDK.
+
+1. **Checkout** 1xx branch:
+2. **Build** using `./build.sh --with-sdk <PSB 1xx SDK> --with-artifacts <PSB 1xx artifacts>`.
+3. **Gather** 1xx build outputs from `artifacts/x64/Release` = `<CSB 1xx artifacts>`.
+4. **Prepare** 2xx inputs - `<PSB 2xx artifacts> + <CSB 1xx artifacts> + <PSB 1xx artifacts>`
+5. **Checkout** 2xx branch
+6. **Build** using `./build.sh --with-sdk <PSB 2xx SDK> --with-artifacts <2xx inputs>`
+
+### A distro maintainer wishing to support all available SDKs (monthly workflow)
+
+A distro maintainer wishing to support all SDK bands builds the 1xx band, then may build all Nxx bands in parallel.
+
+1. **Checkout** 1xx branch:
+2. **Build** using `./build.sh --with-sdk <PSB 1xx SDK> --with-artifacts <PSB 1xx artifacts>`.
+3. **Gather** 1xx build outputs from `artifacts/x64/Release` = `<CSB 1xx artifacts>`.
+
+The following additional steps may be done in parallel:
+
+For each SDK band Nxx:
+
+   1. **Prepare** Nxx inputs - `<PSB Nxx artifacts> + <CSB 1xx artifacts> + <PSB 1xx artifacts>`
+   2. **Checkout** Nxx branch
+   3. **Build** using `./build.sh --with-sdk <PSB Nxx SDK> --with-artifacts <Nxx inputs>`
+
+## Poison and Prebuilt detection
+
+Prebuilt detection should not change. Poisoning workflows may need to change. We don't want to poison inputs from the 1xx VMR, used when building the 2xx VMR. We **do** want to poison the previously source built inputs prior to adding in the 1xx inputs. So, in a poisoning check on a 2xx VMR build, the following workflow would be used:
+
+1. **Checkout** 1xx branch:
+2. **Build** using `./build.sh --with-sdk <PSB 1xx SDK> --with-artifacts <PSB 1xx artifacts>`.
+3. **Gather** 1xx build outputs from `artifacts/x64/Release` = `<CSB 1xx artifacts>`.
+4. **Prepare (pre-poison)** 2xx inputs - `<PSB 2xx artifacts> + <PSB 1xx artifacts>`
+5. **Poison** 2xx inputs
```

Comment:
I would prefer the types are artifacts are distinguishable within the build instead of being combined.  It keeps the poisoning process simpler.

Reply:
I'm not exactly sure what "pre-poisoning" means. Is it to merge prebuilt artifacts from 1xx with prebuilt artifacts of an externally sourced (i.e. tarball on a Microsoft server) build of 2xx? I ask this, as this breaks N+1. I should be able to build the first release of 2xx with the last prebuilt artifacts of 1xx. As a distro maintainer, the only time I should be introducing externally sourced prebuilt artifacts is when I boostrap a new major release of dotnet (i.e. .net 6 or 7) on a a new release of Alpine.

In fact, the Alpine Linux package doesn't even use `prep.sh`, as the tarballs upstream provides don't have libc-musl artifacts. Our `stage0` package builds these libc-musl specific artifacts, and packages them with the architecture agnostic upstream artifacts. 

I also worry of the increased complexity of bootstrapping a new release of Alpine Linux. At Alpine, for every release, we rebuild the whole dotnet stack. We source the Microsoft linux-musl bootstrap and artifacts, we build the arch-specific nupkgs, and then we build 1xx source-build. I should then be able to build 2xx, and then 3xx, and then 4xx only using artifacts that have been built and packaged on Alpine build servers.

Reply:
I updated this when I split the input switches to --with-previous-artifacts and --with-current-artifacts. Basically, we only want to poison the previous artifacts, not the current ones. The current ones may be redistributed in outputs, the previous must not be.

On the building for a new alpine release, yeah I agree it would be ideal to be able to build the Nxx bands with the 1xx build. I think it is **likely** it will work in most cases (and we can document which cases), but we can't make that garuantee. The SDK bands represent a bunch of development work that could require changes to the input build tools before they GA. I'll discuss this with the team some more.